application/common/models/Email.php

```diff
@@ -0,0 +1,284 @@
+
+        /*
+         * Conditionally set optional fields
+         */
+        $setMethods = [
+            'setCc' => $this->cc_address,
+            'setBcc' => $this->bcc_address,
+        ];
+        foreach ($setMethods as $method => $value) {
+            if ($value) {
+                $mailer->$method($value);
+            }
+        }
+
+        return $mailer;
+    }
+
+    /**
+     * Attempt to send messages from queue
+     * @throws Exception
+     */
+    public static function sendQueuedEmail()
+    {
+        $log = [
+            'action' => 'email/sendQueuedEmail',
+        ];
+        try {
+            $batchSize = \Yii::$app->params['emailQueueBatchSize'];
+            $queued = self::find()->orderBy(['updated_at' => SORT_ASC])
+                ->where(['<', 'send_after', time()])->orWhere('send_after IS NULL')
+                ->limit($batchSize)->all();
+
+            $log += [
+                'batchSize' => $batchSize,
+                'queuedEmails' => count($queued),
+                'sentEmails' => 0,
+            ];
+
+            if (empty($queued)) {
+                // If nothing queued, no need to send log
+                return;
+            }
+
+            /** @var Email $email */
+            foreach ($queued as $email) {
+                $log['sentEmails'] += $email->retry();
+            }
+        } catch (Exception $e) {
+            $log += [
+                'status' => 'error',
+                'error' => $e->getMessage(),
+                'code' => $e->getCode(),
+            ];
+            \Yii::error($log);
+        }
+
+        // Send log of successful processing
+        \Yii::info($log);
+    }
+
+    /**
+     * If $this has been saved to database, it will be deleted and on failure throw an exception
+     * @throws Exception
+     */
+    private function removeFromQueue()
+    {
+        try {
+            if ($this->id && !$this->delete()) {
+                throw new Exception(
+                    'Unable to delete email queue entry',
+                    1461012183
+                );
+            }
+        } catch (Exception $e) {
+            $log = [
+                'action' => 'delete after send',
+                'status' => 'failed to delete',
+                'error' => $e->getMessage(),
+            ];
+            \Yii::error($log, 'application');
+
+            throw new Exception(
+                'Unable to delete email queue entry',
+                1461012337
+            );
+        }
+    }
+
+    /**
+     * @return array of fields that should be included in responses.
+     */
+    public function fields(): array
+    {
+        return [
+            'id',
+            'to_address',
+            'cc_address',
+            'bcc_address',
+            'subject',
+            'text_body',
+            'html_body',
+            'attempts_count',
+            'updated_at',
+            'created_at',
+            'error',
+            'send_after',
+        ];
+    }
```

Comment:
This might not be needed any more. I _think_ this `fields()` function is for controlling what is included when an Active Record model is returned by an API controller action (e.g. as JSON in an HTTP response). I'm not sure of the implications of removing it, though, for what that's worth.

Reply:
I may implement the API endpoints so idp-pw-api and/or idp-id-sync can use id-broker as an email-service. In that case, the `fields` method would be used again.

Thoughts?

Reply:
Sounds good. Seems worth keeping it then. 👍